syntax/parse.rs

```diff
@@ -567,7 +567,21 @@ fn parse_extern_fn(
                     });
                     continue;
                 }
-                return Err(Error::new_spanned(arg, "unsupported signature"));
+
+                match lang {
+                    Lang::Rust => {
+                        return Err(Error::new_spanned(
+                            arg,
+                            "cannot taking self by-value, consider using `&self` or `&mut self` instead"
+                        ));
+                    }
+                    Lang::Cxx => {
+                        return Err(Error::new_spanned(
+                            arg,
+                            "cannot taking self by-value, consider using `self: Pin<&mut Self>`, `self: Box<T>` or `self: UniquePtr<T>` instead"
```

Comment:
`self: UniquePtr<T>` is not allowed on stable Rust so this shouldn't suggest it. See https://github.com/rust-lang/rust/issues/44874.

```console
error[E0658]: `UniquePtr<T>` cannot be used as the type of `self` without the `arbitrary_self_types` feature
  --> src/main.rs:11:16
   |
11 |     fn f(self: UniquePtr<T>) {}
   |                ^^^^^^^^^^^^
   |
   = note: see issue #44874 <https://github.com/rust-lang/rust/issues/44874> for more information
   = help: add `#![feature(arbitrary_self_types)]` to the crate attributes to enable
   = help: consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)
```

Separately, `self: Box<T>` is not a thing that C++ is able to do, because it can only call non-static member functions of `T` on `T&`, `T const&`, `T&&`, and `T const&&`.

The way they'd want to transfer ownership is by having a static member function or non-member function where `Box<T>` or `UniquePtr<T>` is the argument type (as opposed to receiver).